### PR TITLE
Story/ccls 2005 org create

### PR DESCRIPTION
**Pull request by porritta**
Includes
https://dsdmoj.atlassian.net/browse/CCLS-2001
https://dsdmoj.atlassian.net/browse/CCLS-2002
https://dsdmoj.atlassian.net/browse/CCLS-2005

### CODE REVIEW

File: src/main/java/uk/gov/laa/ccms/caab/bean/validators/opponent/IndividualOpponentValidator.java [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+    if (StringUtils.hasText(opponentFormData.getCertificateNumber())) {
+      validateFieldFormat("certificateNumber",
+          opponentFormData.getCertificateNumber(),
+          ALPHA_NUMERIC_SLASH_SPACE_STRING,
+          "Certificate number",
+          errors);
+    }
+  }
+
+  private void validateNameComponent(final String fieldId,
```
Comment (PhilDigitalJustice) [inline src/main/java/uk/gov/laa/ccms/caab/bean/validators/opponent/IndividualOpponentValidator.java:97]: jdoc

File: src/main/java/uk/gov/laa/ccms/caab/bean/validators/opponent/IndividualOpponentValidator.java [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+    }
+
+    if (StringUtils.hasText(fieldValue)) {
+      validateFieldFormat(fieldId, fieldValue, characterSetRestriction, displayValue, errors);
+      validateFirstCharAlpha(fieldId, fieldValue, displayValue, errors);
+      validateDoubleSpaces(fieldId, fieldValue, displayValue, errors);
+    }
+  }
+
+  private void validateDateComponent(final String fieldId, final String fieldValue,
```
Comment (PhilDigitalJustice) [inline src/main/java/uk/gov/laa/ccms/caab/bean/validators/opponent/IndividualOpponentValidator.java:114]: jdoc

File: src/main/java/uk/gov/laa/ccms/caab/bean/validators/opponent/OrganisationOpponentValidator.java [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+      validateTelephoneNumber("telephoneWork", opponentFormData.getTelephoneWork(),
+          false, "Telephone", errors);
+      validateTelephoneNumber("faxNumber", opponentFormData.getFaxNumber(),
+          false, "Fax", errors);
+    }
+
+    validateOtherInformation(opponentFormData, errors);
+  }
+
+  private void validateOtherInformation(OrganisationOpponentFormData opponentFormData,
```
Comment (PhilDigitalJustice) [inline src/main/java/uk/gov/laa/ccms/caab/bean/validators/opponent/OrganisationOpponentValidator.java:63]: jdoc

File: src/main/java/uk/gov/laa/ccms/caab/bean/validators/opponent/OrganisationOpponentValidator.java [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+      Errors errors) {
+    if (StringUtils.hasText(opponentFormData.getOtherInformation())) {
+      validateFieldFormat("otherInformation", opponentFormData.getOtherInformation(),
+          STANDARD_CHARACTER_SET, "Other information", errors);
+      validateFieldMaxLength("otherInformation", opponentFormData.getOtherInformation(),
+          2000, "Other information", errors);
+    }
+  }
+
+  private void validateOrganisationName(Errors errors,
```
Comment (PhilDigitalJustice) [inline src/main/java/uk/gov/laa/ccms/caab/bean/validators/opponent/OrganisationOpponentValidator.java:73]: jdoc

File: src/main/java/uk/gov/laa/ccms/caab/bean/validators/opponent/OrganisationOpponentValidator.java [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+    validateRequiredField("organisationName", opponentFormData.getOrganisationName(),
+        "Organisation name", errors);
+
+    if (StringUtils.hasText(opponentFormData.getOrganisationName())) {
+      validateFieldFormat("organisationName", opponentFormData.getOrganisationName(),
+          CHARACTER_SET_A, "Organisation name", errors);
+    }
+  }
+
+  protected void validateContactNameRole(final String contactNameRole, final Errors errors) {
```
Comment (PhilDigitalJustice) [inline src/main/java/uk/gov/laa/ccms/caab/bean/validators/opponent/OrganisationOpponentValidator.java:84]: jdoc

File: src/main/java/uk/gov/laa/ccms/caab/controller/application/summary/OpponentsSectionController.java [hunk 6]
```diff
@@ -275,4 +554,49 @@ private void populateConfirmSharedOrganisationDropdowns(final Model model) {
...
+    model.addAttribute("relationshipsToClient", combinedLookup.getT3().getContent());
+    model.addAttribute("countries", combinedLookup.getT4().getContent());
+
+    List<Pair<Boolean, String>> currentlyTradingOptions = List.of(
+        Pair.of(Boolean.FALSE, "No"),
+        Pair.of(Boolean.TRUE, "Yes"));
+    model.addAttribute("currentlyTradingOptions", currentlyTradingOptions);
+  }
+
+  private void populateIndividualCreateDropdowns(final Model model) {
```
Comment (PhilDigitalJustice) [inline src/main/java/uk/gov/laa/ccms/caab/controller/application/summary/OpponentsSectionController.java:580]: if common lookups can we use the dropdownBuilder?

File: src/main/java/uk/gov/laa/ccms/caab/controller/application/summary/OpponentsSectionController.java [hunk 7]
```diff
@@ -275,4 +554,49 @@ private void populateConfirmSharedOrganisationDropdowns(final Model model) {
     model.addAttribute("relationshipsToClient", combinedLookup.getT2().getContent());
   }
 
+  private void populateOrganisationCreateDropdowns(final Model model) {
```
Comment (PhilDigitalJustice) [inline src/main/java/uk/gov/laa/ccms/caab/controller/application/summary/OpponentsSectionController.java:557]: if common lookups can we use the dropdownBuilder?

File: src/main/resources/templates/application/opponents/opponents-remove.html [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+  <div id="breadcrumbs">
+    <a th:href="@{/application/summary/opponents}" class="govuk-back-link">Cancel and return to opponents and other parties</a>
+  </div>
+  <div id="main-content">
+    <form action="#" th:action="@{/application/opponents/{opponentId}/remove(opponentId=${opponentId})}" method="post">
+      <h1 class="govuk-heading-l">
+        Delete opponent
+      </h1>
+      <p class="govuk-body-l">
+        Are you sure you want to delete this opponent? Please click 'Confirm' to delete
```
Comment (PhilDigitalJustice) [inline src/main/resources/templates/application/opponents/opponents-remove.html:17]: Do you need to specify the opponent we are deleting?

File: src/main/resources/templates/application/partials/opponent-fragments.html [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+    <div th:replace="~{partials/forms :: textInput('addressLine2', 'Address line 2 (optional)', 'govuk-input--width-20')}"></div>
+    <div th:replace="~{partials/forms :: textInput('city', 'City / town (optional)', 'govuk-input--width-20')}"></div>
+    <div th:replace="~{partials/forms :: textInput('county', 'County (optional)', 'govuk-input--width-20')}"></div>
+    <div th:replace="~{partials/forms :: dropdown('country', 'Country (optional)', ${countries}, 'code', 'description', false)}"></div>
+    <div th:replace="~{partials/forms :: textInput('postcode', 'Post code (optional)', 'govuk-input--width-20')}"></div>
+    <div th:replace="~{partials/forms :: textInput('telephoneWork', 'Telephone (optional)', 'govuk-input--width-20')}"></div>
+    <div th:replace="~{partials/forms :: textInput('faxNumber', 'Fax (optional)', 'govuk-input--width-20')}"></div>
+    <div th:replace="~{partials/forms :: textInput('emailAddress', 'Email address (optional)', 'govuk-input--width-20')}"></div>
+    <div th:replace="~{partials/forms :: largeTextInput('otherInformation', 'Any other information about the liable organisation  (Maximum 2000 characters allowed)')}"></div>
```
Comment (PhilDigitalJustice) [inline src/main/resources/templates/application/partials/opponent-fragments.html:53]: do we actually limit the form to only allow the 2000 characters?

**Review by PhilDigitalJustice — CHANGES_REQUESTED** [review]
A few comments to resolve

File: src/main/resources/templates/application/partials/opponent-fragments.html (hunk 9 above)
Comment (porritta) [reply src/main/resources/templates/application/partials/opponent-fragments.html:53]: This is handled by the form validator on POST

File: src/main/java/uk/gov/laa/ccms/caab/bean/validators/opponent/IndividualOpponentValidator.java (hunk 2 above)
Comment (porritta) [reply src/main/java/uk/gov/laa/ccms/caab/bean/validators/opponent/IndividualOpponentValidator.java:138]: added

File: src/main/java/uk/gov/laa/ccms/caab/bean/validators/opponent/OrganisationOpponentValidator.java (hunk 3 above)
Comment (porritta) [reply src/main/java/uk/gov/laa/ccms/caab/bean/validators/opponent/OrganisationOpponentValidator.java]: added

File: src/main/java/uk/gov/laa/ccms/caab/bean/validators/opponent/OrganisationOpponentValidator.java (hunk 4 above)
Comment (porritta) [reply src/main/java/uk/gov/laa/ccms/caab/bean/validators/opponent/OrganisationOpponentValidator.java]: added

File: src/main/java/uk/gov/laa/ccms/caab/bean/validators/opponent/OrganisationOpponentValidator.java (hunk 5 above)
Comment (porritta) [reply src/main/java/uk/gov/laa/ccms/caab/bean/validators/opponent/OrganisationOpponentValidator.java]: added

File: src/main/java/uk/gov/laa/ccms/caab/controller/application/summary/OpponentsSectionController.java (hunk 6 above)
Comment (porritta) [reply src/main/java/uk/gov/laa/ccms/caab/controller/application/summary/OpponentsSectionController.java:577]: done

File: src/main/java/uk/gov/laa/ccms/caab/controller/application/summary/OpponentsSectionController.java (hunk 7 above)
Comment (porritta) [reply src/main/java/uk/gov/laa/ccms/caab/controller/application/summary/OpponentsSectionController.java:554]: done

File: src/main/resources/templates/application/opponents/opponents-remove.html (hunk 8 above)
Comment (porritta) [reply src/main/resources/templates/application/opponents/opponents-remove.html]: opponent party name added to view

**Review by PhilDigitalJustice — APPROVED** [review]

**Comment by PhilDigitalJustice** [comment]
LGTM

**Pull request merged**